clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java

```diff
@@ -86,6 +85,25 @@ public AbortedTransaction(long pid, long firstOffset) {
         }
 
         @Override
+        public boolean equals(Object o) {
+            if (this == o)
+                return true;
+            if (o == null || getClass() != o.getClass())
+                return false;
+
+            AbortedTransaction that = (AbortedTransaction) o;
+
+            return pid == that.pid && firstOffset == that.firstOffset;
+        }
+
+        @Override
+        public int hashCode() {
+            int result = (int) (pid ^ (pid >>> 32));
```

Comment:
Also, yes, it's old-fashioned to have to implement these methods. Unfortunately, Java will only have data classes in Java 10/11 (they are somewhat similar to Scala case classes).